Integrations/python/deephaven/learn/gather/__init__.py

```diff
@@ -0,0 +1,122 @@
+# 
+# Copyright (c) 2016 - 2021 Deephaven Data Labs and Patent Pending
+# 
+"""
+Utilities for gathering Deephaven table data into Python objects
+"""
+
+import numpy as np
+import jpy
+import wrapt
+
+# None until the first _defineSymbols() call
+_gatherer = None
+
+def _defineSymbols():
+    if not jpy.has_jvm():
+        raise SystemError("No java functionality can be used until the JVM has been initialized through the jpy module")
+
+    global _gatherer
+
+    if _gatherer is None:
+        _gatherer = jpy.get_type("io.deephaven.integrations.learn.Gatherer")
+
+# Every method that depends on symbols defined via _defineSymbols() should be decorated with @_passThrough
+@wrapt.decorator
+def _passThrough(wrapped, instance, args, kwargs):
+    """
+    For decoration of module methods, to define necessary symbols at runtime
+
+    :param wrapped: the method to be decorated
+    :param instance: the object to which the wrapped function was bound when it was called
+    :param args: the argument list for `wrapped`
+    :param kwargs: the keyword argument dictionary for `wrapped`
+    :return: the decorated version of the method
+    """
+
+    _defineSymbols()
+    return wrapped(*args, **kwargs)
+
+try:
+    _defineSymbols()
+except Exception as e:
+    pass
+
+@_passThrough
+def table_to_numpy_2d(row_set, col_set, order = 0, dtype = None):
+    """
+    Convert Deephaven table data to a 2d NumPy array of the appropriate size
+
+    :param row_set: A RowSequence describing the number of rows in the table
+    :param col_set: ColumnSources describing which columns to copy
+    :param order: The major order of copying -> either row major or column major
+    :param dtype: The desired NumPy data type of the output NumPy array
+    :return: A NumPy ndarray
+    """
+
+    if dtype == bool:
+        dtype = np.bool_
+    elif dtype == float:
+        dtype = np.double
+    elif dtype == int:
+        dtype = np.intc
+
+    if dtype == np.bool_:
+        if order == 0:
+            buffer = _gatherer.tensorBuffer2DBooleanColumns(row_set, col_set)
+        elif order == 1:
+            buffer = _gatherer.tensorBuffer2DBooleanRows(row_set, col_set)
```

Comment:
After some thinking about this, and the `order` constant, I think you can massively simplify things by:
1) Creating a Java function called `tensorBuffer2DBoolean` with a bool argument for row vs column major.  This function would then call either `tensorBuffer2DBooleanColumnMajor` or `tensorBuffer2DBooleanRowMajor`.

2) Use a Python enum with values to for `order`.  This link below gives some details.
https://docs.python.org/3/library/enum.html#:~:text=Planet-,%C2%B6,-If%20__new__()%20or

It would look something like:
```
class MemoryLayout(Enum):
    ROW_MAJOR = (True)
    COLUMN_MAJOR = (False)
    C = (True) 
    FORTRAN = (False)

    def __init(self, is_row_major)
        self.is_row_major = is_row_major

def table_to_numpy_2d(row_set, col_set, order:MemoryLayout=MemoryLayout.ROW_MAJOR, dtype:numpy.dtype = numpy.float64) -> numpy.ndarray
```